### PR TITLE
Fixes and miscs

### PR DESCRIPTION
**Pull request by pent0**
- Fix Flood It compile error: error C7011: implicit cast from "vec4" to "vec2", and some bugs in optimization and hacks
- A part of TODO: Correct secondary attribute (SECATTR) generation
- Add support for generating and access array types
- Implement VMIN, VMAX instructions
- Remove unroll

The reason for i say a part is: this doesn't not correct sampler SA offset. I don't know how they are generated yet.

### CODE REVIEW

**Comment by Zangetsu38** [comment]
Without last two commit, with shader folder, on HB sdlpal
[vita3k.log](https://github.com/Vita3K/Vita3K/files/2934241/vita3k.log)

**Comment by pent0** [comment]
Failed shader compilations should be fixed in f4a8450

**Comment by pent0** [comment]
Good to review i suppose

**Comment by petmac** [comment]
Looks good to me, but probable needs review from someone more familiar with it.

**Comment by VelocityRa** [comment]
Yeah I'll review tomorrow probably.

**Comment by pent0** [comment]
Take your time :D

**Comment by Zangetsu38** [comment]
Problem for lot game without shader folder, it is regression compared master, for Cave Story, retroArch nxengine, VitaHexenII, NumptyPhysics.

File: src/emulator/shader/src/usse_translator.cpp [hunk 1]
```diff
@@ -273,20 +273,18 @@ spv::Id USSETranslatorVisitor::load(Operand &op, const Imm4 dest_mask, const std
...
-                return m_b.createLoad(reg_left.var_id);
-            else if (is_reg_right_comp_offset)
-                return m_b.createLoad(reg_right.var_id);
-        }
+        const uint32_t reg_left_comp_count = m_b.getNumTypeComponents(reg_left.type_id);
+        const uint32_t reg_right_comp_count = m_b.getNumTypeComponents(reg_right.type_id);
+
+        const bool is_reg_left_comp_offset = (reg_left_comp_offset == 0);
+        const bool is_reg_left_equal_dest_comp_count = (reg_left_comp_count == dest_comp_count);
+        const bool is_reg_right_comp_offset = (reg_right_comp_offset == dest_comp_count);
```
Comment (pent0) [inline src/emulator/shader/src/usse_translator.cpp]: This line merely works by accident, will rework on it

**Comment by Zangetsu38** [comment]
good retroarch, HexenII and Numpty physics fixed

**Comment by Zangetsu38** [comment]
Other regression founded on Amnesia: Memories game, with your PR, game crash during try create few shaders, works no problem on master, log with your PR, you can see in end what shaders causing first crash
[vita3k.log](https://github.com/Vita3K/Vita3K/files/2950575/vita3k.log)
you can compare with master here
[vita3k.log](https://github.com/Vita3K/Vita3K/files/2950577/vita3k.log)

**Comment by VelocityRa** [comment]
@pent0 Should I review this?

**Comment by pent0** [comment]
@VelocityRa it's ready

**Pull request merged**